Support Pod-Resources in crd and statefulset

christianwoehrle opened the pull request:
## JIRA ID
KEYCLOAK-12681

## Additional Information

I want to specify these requests and limits for all my pods in the cluster.

The crd ``` keycloaks.keycloak.org ``` is extended to support that for keycloak and for keycloak-postresql.

If requests are not set in the custom-resource, the operator provides a sane default for keycloak-memory and for keycloak-postgres memory and cpu.

## Verification Steps

Create a new keycloak cr without resources set:
```
kubectl apply -f -
apiVersion: keycloak.org/v1alpha1
kind: Keycloak
metadata:
  name: keycloak
  labels:
    app: sso
    test: sso
spec:
  instances: 1
  extensions:
    - https://github.com/aerogear/keycloak-metrics-spi/releases/download/1.0.4/keycloak-metrics-spi-1.0.4.jar
  externalAccess:
    enabled: False
  podDisruptionBudget:
    enabled: True
```
Check that keycloak and keycloak-postgres have been created without resources
```
kubectl get statefulsets.apps keycloak  -o jsonpath="{ ..resources }"  
map[] map[]✔ 

kubectl get deployment keycloak-postgresql  -o jsonpath="{ ..resources }"  
map[]✔ 
```



Create/Update the keycloak-cr with resources.
```
kubectl apply -f -
apiVersion: keycloak.org/v1alpha1
kind: Keycloak
metadata:
  name: keycloak
  labels:
    app: sso
    test: sso
spec:
  instances: 1
  extensions:
    - https://github.com/aerogear/keycloak-metrics-spi/releases/download/1.0.4/keycloak-metrics-spi-1.0.4.jar
  externalAccess:
    enabled: False
  podDisruptionBudget:
    enabled: True
  keycloakDeploymentSpec:
    resources:
      requests:
        cpu: 2500m
        memory: 700Mi
  postgresDeploymentSpec:
    resources:
      requests:
        cpu: 100m
        memory: 150Mi

```

Check the results (give it some time)
```
kubectl get statefulsets.apps keycloak  -o jsonpath="{ ..resources }"  
map[] map[requests:map[cpu:2500m memory:700Mi]]✔

kubectl get deployment keycloak-postgresql  -o jsonpath="{ ..resources }"  
map[requests:map[cpu:100m memory:150Mi]]✔```
```


Create/Update the keycloak-cr with resources and limits.

```
kubectl apply -f -
apiVersion: keycloak.org/v1alpha1
kind: Keycloak
metadata:
  name: keycloak
  labels:
    app: sso
    test: sso
spec:
  instances: 1
  extensions:
    - https://github.com/aerogear/keycloak-metrics-spi/releases/download/1.0.4/keycloak-metrics-spi-1.0.4.jar
  externalAccess:
    enabled: False
  podDisruptionBudget:
    enabled: True
  keycloakDeploymentSpec:
    resources:
      requests:
        cpu: 2700m
        memory: 800Mi
      limits:
        cpu: 2700m
        memory: 800Mi
  postgresDeploymentSpec:
    resources:
      requests:
        cpu: 250m
        memory: 250Mi
      limits:
        cpu: 250m
        memory: 250Mi
```

Check that the change has happened (give it some time)
```
kubectl get statefulsets.apps keycloak  -o jsonpath="{ ..resources }"  
map[] map[limits:map[cpu:2700m memory:800Mi] requests:map[cpu:2700m memory:800Mi]]

kubectl get deployment keycloak-postgresql  -o jsonpath="{ ..resources }"  
map[limits:map[cpu:250m memory:250Mi] requests:map[cpu:250m memory:250Mi]]✔ 
```

## Checklist:
- [ ] Verified by team member
- [ ] Comments where necessary
- [ ] Automated Tests
- [ ] Documentation changes if necessary

## Additional Notes
<!-- PS.: Add images and/or .gifs to illustrate what was changed if this pull request modifies the appearance/output of something presented to the users. -->

kfox1111 commented:
Big +1 on this use case. Being able to specify resources will be very important to running in production.

r3kzi commented:
+1

Without this we won't be running the operator in our cluster.

christianwoehrle commented:
Further discussed in https://issues.redhat.com/browse/KEYCLOAK-12681

kfox1111 commented:
Any movement?

slaskawi commented on pkg/apis/keycloak/v1alpha1/keycloak_types.go:
```diff
@@ -51,6 +51,36 @@ type KeycloakSpec struct {
+	PostgresDeploymentSpec PostgresDeploymentSpec `json:"postgresDeploymentSpec,omitempty"`
+}
+
+type KeycloakDeploymentSpec struct {
```
I believe we could have a single struct for both Keycloak and Postgresql (comments and annotations omitted). Something like:

```
type DeploymentSpec struct {
	KeycloakDeploymentSpec ResourcesAndLimitsSpec `json:"keycloakDeploymentSpec,omitempty"`
	PostgresDeploymentSpec ResourcesAndLimitsSpec `json:"postgresDeploymentSpec,omitempty"`
}

type ResourcesAndLimitsSpec struct {
	Limits   ResourceList `json:"limits,omitempty"`
	Requests ResourceList `json:"requests,omitempty"`
}
```

The `ResourceList` might be taken from Kubernetes `v1` package (see` k8s.io/api/core/v1/types.go`). Implementing it this way, will save us some lines of code required for converting types from one to another.

christianwoehrle replied:
Single struct sounds good, this is probably too much. And using ResourceList-type from k8s sounds right too.

What do you think of this:
KeycloakSpec has now two separate entries for Keycloak- and Postgres-DeploymentSpec, so the structure is less deep.
Both entries have the same Type `DeploymentSpec` which contains the k8s-Type  `ResourceRequirements` with `Limits` and `Requests`.
If someone needs to specify more than just requests and limits, then `DeploymentSpec` can be extended


```
type KeycloakSpec struct {
...
	KeycloakDeploymentSpec DeploymentSpec `json:"keycloakDeploymentSpec,omitempty"`
	PostgresDeploymentSpec DeploymentSpec `json:"postgresDeploymentSpec,omitempty"`
}

type DeploymentSpec struct {
	// Resources (Requests and Limits) for the KeycloakPods
	// +optional
	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
}

slaskawi commented on pkg/model/keycloak_deployment.go:
```diff
@@ -19,12 +21,56 @@ const (
 )
 
+const (
+	KeycloakPodDefaultMemoryRequests = "600Mi"
```
Let's remove the default. As you pointed out in the JIRA (or Keycloak-dev, I don't quite remember) the default might be different depending on the Providers you're using.

christianwoehrle replied:
ok

christianwoehrle commented:
Thank you for the review.
I added some tests, removed the defaults, updated linter errors in the files contained in this merge requests, and cleaned up the unnecessary types in keycloak_types.go.

There is still a build error from test/e2e because of the missing kubeconfig. The same error is in other PRs too, so it looks like this is not a result of this PR.
The remaining lint errors are from different that I haven't changed in this PR.

Please let me know if there's something missing.

slaskawi commented:
Thank for addressing the comments. I'll have a look at this PR early next week. Please stay tuned.

slaskawi commented:
Any final words, @mhajas ? If not, I plan to merge it early next week.

slaskawi commented:
Thanks @christianwoehrle !
